### PR TITLE
Setup Models, Routes, and Importer

### PR DESCRIPTION
**Pull request by michellemtchai**
- add `Applicant`, `Instructor` and `Position` from TAPP as ActiveResource
- create `Offer` and `Contract` model
- updated `dev.env.default` because it was using the tapp setting and the server couldn't run as a result
- setup routes in `localhost:3001`
  - `GET /applicants`
  - `GET /applicants/:id`
  - `GET /offers`
  - `GET /offers/:id`
  - `PATCH /offers/:id` for attribute `objection`
  - `POST /offers/:id/send-contract` --> incomplete: waiting for mailer function
  - `GET /offers/instructor/:instructor_id`
  - `GET /contracts`
  - `GET /contracts/:id`
  - `PATCH /contracts/:id` for attributes `withdrawn`, `accepted`, `printed`
  - `POST /contracts/:id/nag` --> incomplete: waiting for mailer function
- to run the app do the following in order:
  - go to the TAPP repo and run `docker-compose up`
  - go to this repo and run `cp dev.env.default .env` then `docker-compose up`
  - run `docker-compose run rails-app rake db:migrate db:seed`
- created `OfferImporter` based on guesses
- finish RSpec for the `ApplicantsController`, `ContractsController` and `OffersController`

Explanation:
- run `docker-compose run rails-app rake db:migrate db:seed` to get data into system

Note:
- This PR has been modified using the CP to TAPP connection based on PR#3 (CP) and PR#110 (TAPP)

### CODE REVIEW

File: db/seeds/mock_offers.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
+{
```
Comment (jmzaleski) [inline db/seeds/mock_offers.json]: wow michelle, where did you get the format from? 
I was thinking we had to drive our CHASS development image to the point of exporting offers!

File: dev.env.default [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Environment variables needed for running this project
 # copy this file to .env in this directory for it to be picked up by docker-compose
-POSTGRES_DB=tapp_development
-POSTGRES_USER=tapp
+POSTGRES_DB=cp_development
```
Comment (jmzaleski) [inline dev.env.default]: tapp and cp are separate postgress dbs.. do you think it would be okay to leave the users the same?

File: run_cp.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+cp dev.env.default .env
+temp="`docker inspect -f '{{range .NetworkSettings.Networks}}{{.Gateway}}{{end}}' tapp_rails-app_1`"
```
Comment (jmzaleski) [inline run_cp.sh]: yikes! this is heroic, but probably only will serve to make prototype the endpoint access from cp to tapp. (It makes me  wonder if keeping the images separate is worth the kind of hypothetical security benefit)

File: db/seeds/mock_offers.json (hunk 1 above)
Comment (michellemtchai) [inline db/seeds/mock_offers.json]: Like I mentioned in the description, this is mostly just a guess. This isn't the official format.

File: Dockerfile [hunk 4]
```diff
@@ -23,5 +23,5 @@ RUN yarn install
 
 COPY . /srv/app/
 
-EXPOSE 3000
-CMD ["rails", "server", "-b", "0.0.0.0", "-p", "3000"]
+EXPOSE 3001
+CMD ["rails", "server", "-b", "0.0.0.0", "-p", "3001"]
```
Comment (freeatnet) [inline Dockerfile]: I don't think we should change this port: since tapp and cp rails apps run in separate containers, using port 3000 on both sides would not conflict.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,3 +1,11 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  rescue_from (Errno::ECONNREFUSED), with: :record_not_found
+  rescue_from ActiveRecord::RecordNotFound, with: :record_not_found
+  rescue_from StandardError, with: :record_not_found
```
Comment (freeatnet) [inline app/controllers/application_controller.rb]: `Errno::ECONNREFUSED` and `StandardError` should probably not be resulting in a 404 — and definitely should not be swallowed quietly by the handler.

File: app/controllers/contracts_controller.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
+class ContractsController < ApplicationController
+  protect_from_forgery with: :null_session
+
+  def index
+    contracts = []
+    Contract.all.each do |contract|
+      contracts.push(format_contract(contract))
+    end
+    render json: contracts
```
Comment (freeatnet) [inline app/controllers/contracts_controller.rb]: `Contract.all.includes(:offer).map { |c| format_contract(c) }` would be more succinct.

File: app/controllers/contracts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+  def update
+    contract = Contract.find(params[:id])
+    contract.update_attributes!(contract_params)
+  end
+
+  def nag
+    contract = Contract.find(params[:id])
+    # send out email remainders
+    contract.update_attributes!(nag_count: contract[:nag_count]+1)
```
Comment (freeatnet) [inline app/controllers/contracts_controller.rb]: We could use [`increment`/`increment!`](http://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-increment) here.

File: app/controllers/contracts_controller.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+  def format_contract(contract)
+    offer = contract.offer
+    contract = contract.as_json
+    position = Position.find(offer[:position_id]).as_json
+    contract["position"] = position["position"]
+    applicant = Applicant.find(offer[:applicant_id]).as_json
+    contract["applicant"] = applicant
+    contract["deadline"] = (contract["created_at"] + ENV["deadline"].to_i)
+    contract["withdrawn"] = Time.now > contract["deadline"]
```
Comment (freeatnet) [inline app/controllers/contracts_controller.rb]: `contract.merge(…hash of attributes…)` might be better — you could also then omit `return contract`

File: app/controllers/offers_controller.rb [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  end
+
+  private
+  def offer_params
+    params.permit(:objection)
+  end
+
+  def get_all_offers
+    offers = []
+    Offer.all.each do |offer|
```
Comment (freeatnet) [inline app/controllers/offers_controller.rb]: Same deal with `map` over `all.each`

File: app/controllers/offers_controller.rb [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+  def show
+    offer = Offer.find(params[:id])
+    render json: format_offer(offer)
+  end
+
+  def show_by_instructor
+    offers = []
+    get_all_offers.each do |offer|
+      if taught_by(offer, params[:instructor_id].to_i)
```
Comment (freeatnet) [inline app/controllers/offers_controller.rb]: You could use `get_all_offers.filter { |o| taught_by(offer, params[:instructor_id].to_i) }` to obtain an array of offers matching a condition

File: app/controllers/offers_controller.rb [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+  def update
+    offer = Offer.find(params[:id])
+    offer.update_attributes!(offer_params)
+  end
+
+  def send_contract
+    offer = Offer.find(params[:id])
+    if !offer.contract
+      offer.create_contract!(link: "my secret")
```
Comment (freeatnet) [inline app/controllers/offers_controller.rb]: Hmm?

File: app/controllers/offers_controller.rb [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+      if instructor["id"] == instructor_id
+        found = true
+      end
+    end
+    return found
+  end
+
+  def format_offer(offer_rec)
+    offer = offer_rec.as_json
+    offer["sent"] = (offer_rec.contract)? true:false
```
Comment (freeatnet) [inline app/controllers/offers_controller.rb]: Looks like you are checking if `offer_rec.contract` is `nil` here. You could replace the ternary with `offer_rec.contract.present?`.

File: app/models/applicant.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
+require 'active_resource'
```
Comment (freeatnet) [inline app/models/applicant.rb]: Do we really need to explicitly require it here?

File: app/models/applicant.rb [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+require 'net/http'
+class Applicant < ActiveResource::Base
+  self.site = "http://#{ENV['TAPP']}:3000/"
+
+  def self.find_by(utorid)
```
Comment (freeatnet) [inline app/models/applicant.rb]: Should this be called `find` or `find_by_utorid` instead?

File: app/models/applicant.rb [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+    Applicant.all.as_json.each do |applicant|
```
Comment (freeatnet) [inline app/models/applicant.rb]: Check out [`Array.find`](https://ruby-doc.org/core-2.4.1/Enumerable.html#method-i-find).

File: app/models/offer.rb [hunk 16]
```diff
@@ -0,0 +1,4 @@
+class Offer < ApplicationRecord
+  validates_uniqueness_of :position_id, :scope => [:applicant_id]
```
Comment (freeatnet) [inline app/models/offer.rb]: Hash syntax

File: config/application.rb [hunk 17]
```diff
@@ -21,7 +21,8 @@ module MyTAPP
   class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 5.1
-
+    # Set deadline for contract acceptance as 2 weeks
+    ENV["deadline"] = (2*7*24*60*60).to_s
```
Comment (freeatnet) [inline config/application.rb]: Hmm. This should probably be a constant in one of the model classes with a fallback default set there, too.

File: config/puma.rb [hunk 18]
```diff
@@ -9,7 +9,7 @@
 
 # Specifies the `port` that Puma will listen on to receive requests; default is 3000.
 #
-port        ENV.fetch("PORT") { 3000 }
+port        ENV.fetch("PORT") { 5000 }
```
Comment (freeatnet) [inline config/puma.rb]: Hm?

File: config/routes.rb [hunk 19]
```diff
@@ -1,4 +1,11 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   get "/hello_react", to: "hello_react#contracts"
+
+  resources :applicants
+  resources :offers
+  resources :contracts
+  get "offers/instructor/:instructor_id" => "offers#show_by_instructor"
+  post "offers/:id/send-contract" => "offers#send_contract"
+  post "contracts/:id/nag" => "contracts#nag"
```
Comment (freeatnet) [inline config/routes.rb]: You can [nest these](http://guides.rubyonrails.org/routing.html#adding-more-restful-actions) under `resources`

File: package.json [hunk 20]
```diff
@@ -2,6 +2,7 @@
   "name": "MyTAPP",
   "private": true,
   "dependencies": {
+    "activeresource": "^2.0.11",
```
Comment (freeatnet) [inline package.json]: What's this thingy for?

Comment (michellemtchai) [inline package.json]: This was added automatically when I installed the activeresource package.

Comment (freeatnet) [inline package.json]: This doesn't seem to be related to the Ruby activeresource library: https://www.npmjs.com/package/activeresource

File: app/controllers/offers_controller.rb [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    offer = Offer.find(params[:id])
+    offer.update_attributes!(offer_params)
+  end
+
+  def send_contract
+    puts params
+    offer = Offer.find(params[:offer_id])
+    puts offer
+    if !offer.contract
+      offer.create_contract!(link: "mangled-link-for-accepting-offer")
```
Comment (freeatnet) [inline app/controllers/offers_controller.rb]: Awesome!

File: app/models/contract.rb [hunk 22]
```diff
@@ -0,0 +1,7 @@
+class Contract < ApplicationRecord
+  belongs_to :offer
+
+  def get_deadline
+    return self[:created_at] + (2*7*24*60*60)
```
Comment (freeatnet) [inline app/models/contract.rb]: No need for explicit `return` when the value we want to return is the expression on the last line of the function

File: app/controllers/contracts_controller.rb [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  private
+  def contract_params
+    params.permit(:accepted, :printed)
+  end
+
+  def format_contract(contract)
+    offer = contract.offer
+    deadline = (contract.get_deadline)
```
Comment (freeatnet) [inline app/controllers/contracts_controller.rb]: Unnecessary parenthesis.

File: config/routes.rb [hunk 24]
```diff
@@ -1,4 +1,15 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   get "/hello_react", to: "hello_react#contracts"
+
+  resources :applicants
+  resources :offers do
+    post "send-contract" => "offers#send_contract"
+  end
+  resources :contracts do
+    post "nag" => "contracts#nag"
```
Comment (freeatnet) [inline config/routes.rb]: Perfect.
